In [ ]:
#vibo: Vitaliy Bocharov, telegram(@VitaliyBocharov)

![](ods_stickers.jpg)

#### Открытый курс машинного обучения (ведет - Петр Ермаков)
## 003 Введение в python (задание)

In [2]:
#vibo: 003 Введение в python https://youtu.be/rWxD7vZdldY

### Вопрос 1.

Написать код, который вычисляет сумму всех чисел, удовлетворяющих следующие условия:
- положительные целые числа от 1 до 1_000_000_002 (один миллиард два) включительно
- которые нацело (без остатка) делятся на 3 (пример: 3, 6, 9, ...)
- и которые не заканчиваются на 4 и 7 (пример заканчивающихся на 4 и 7: 24, 27, 54, 57 ...)

Результат суммы ввести в поле ниже.
В поле нужно ввести только число, являющееся ответом, без пробелов и других символов.

#### Подготовка

##### Решение-1: brut force - 2min 45s

In [1]:
%%time
#vibo: вариант 1 brut force

#vibo: искомое значение a
a = 0
#vibo: проходим циклом по значениям от 1 до 1 000 000 003 (последнее значение не включается)
for i in range(1, 1000000003):
    #print('i=', i)
    #vibo: три условия (целочисленное даление без остатка на 3;
    #vibo: остаток от целочисленного деления на 10 не должен равняться 4 и 7)
    if i%3==0 and i%10!=4 and i%10!=7:
    #vibo: или так
    #if i%3==0 and not (i%10==4 or i%10==7):
        #print('i=', i)
        #vibo: сумма нарaстающим итогом
        a +=i
        #print('a=', a)
        #print('***')
print('Ответ:', a)

Ответ: 133333334466666672
CPU times: user 2min 45s, sys: 17 ms, total: 2min 45s
Wall time: 2min 45s


##### Решение-2: brut force + генератор - 1min 20s

In [1]:
%%time
#vibo: вариант 2 brut force с генератором чисел кратных трем
a = 0
#vibo: задаем генератор чисел кратных трем
#vibo: сразу значительно сокращаем количество цифр перебора
for i in range(3, 1000000003, 3):
    #vibo: уже проверяем только по двум условиям
    if i%10!=4 and i%10!=7:
        a +=i
print('Ответ:', a)

Ответ: 133333334466666672
CPU times: user 1min 19s, sys: 72.6 ms, total: 1min 19s
Wall time: 1min 20s


##### Решение-: вычитание списков - out of memory

In [2]:
#vibo: только списками вычислить не получается, нехватка памяти
#vibo: пришлось уменьшить значение до 1003, не подходит
a3 = [i for i in range(3, 1003, 3)]
print(a3)
a24 = [i for i in range(24, 1003, 30)]
print(a24)
a27 = [i for i in range(27, 1003, 30)]
print(a27)
#vibo: избавляемся от ненужных значений
#vibo: преобразование в set оставляет только уникальные значения
a_tot = list(set(a3) - set(a24) - set(a27))
print(a_tot)
#vibo: для суммирования нужно от set перейти снова к list
print(sum(a_tot))

[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150, 153, 156, 159, 162, 165, 168, 171, 174, 177, 180, 183, 186, 189, 192, 195, 198, 201, 204, 207, 210, 213, 216, 219, 222, 225, 228, 231, 234, 237, 240, 243, 246, 249, 252, 255, 258, 261, 264, 267, 270, 273, 276, 279, 282, 285, 288, 291, 294, 297, 300, 303, 306, 309, 312, 315, 318, 321, 324, 327, 330, 333, 336, 339, 342, 345, 348, 351, 354, 357, 360, 363, 366, 369, 372, 375, 378, 381, 384, 387, 390, 393, 396, 399, 402, 405, 408, 411, 414, 417, 420, 423, 426, 429, 432, 435, 438, 441, 444, 447, 450, 453, 456, 459, 462, 465, 468, 471, 474, 477, 480, 483, 486, 489, 492, 495, 498, 501, 504, 507, 510, 513, 516, 519, 522, 525, 528, 531, 534, 537, 540, 543, 546, 549, 552, 555, 558, 561, 564, 567, 570, 573, 576, 579, 582, 585, 588, 591, 594, 597, 600, 603, 606, 609, 612, 615, 618, 621, 

##### Решение-3: списки + матидея - 23.1 s

In [4]:
%%time
#vibo: вариант 3 списки и идея арифметической прогрессии

#vibo: задаем список кратный трем, суммируем члены
v3_s_ap1 = sum(i for i in range(3, 1000000003, 3))
print(v3_s_ap1)

#vibo: задаем список 24, 54, и т.д. с шагом +30, суммируем члены
v3_s_ap2 = sum(i for i in range(24, 1000000003, 30))
print(v3_s_ap2)

#vibo: задаем список 27, 57 и т.д. с шагом +30, суммируем члены
v3_s_ap3 = sum(i for i in range(27, 1000000003, 30))
print(v3_s_ap3)

#vibo: из суммы членов первой арифметической прогрессии вычитаем суммы двух других
print('Ответ', v3_s_ap1 - v3_s_ap2 - v3_s_ap3)

166666667833333335
16666666633333332
16666666733333331
Ответ 133333334466666672
CPU times: user 23 s, sys: 3.28 ms, total: 23 s
Wall time: 23.1 s


##### Решение-4: аналитическое решение - 2.89 ms

In [5]:
%%time
#vibo: вараиант 4 аналитическое решение (арифметическая прогрессия)

#vibo: последние члены арифметической прогрессии 24, 54 +30
ap2_an = list((i for i in range(24, 1000000003, 30)))[-1]
print(ap2_an)
#vibo: последние члены арифметической прогрессии 27, 57 +30
ap3_an = list((i for i in range(27, 1000000003, 30)))[-1]
print(ap3_an)

999999984
999999987
CPU times: user 4.4 s, sys: 1.19 s, total: 5.59 s
Wall time: 5.7 s


In [7]:
%%time
#vibo: сумма членов арифметической прогрессии кратных 3
#vibo: первый член
ap1_a1 = 3
#vibo: последний член
ap1_an = 1000000002
#vibo: количество членов (1000000002/3)
ap1_n = 333333334
#vibo: сумма членов армифметической прогрессии
v4_s_ap1 = (ap1_a1 + ap1_an) * ap1_n / 2
print(v4_s_ap1)
print('Проверка v3_s_ap1 - v4_s_ap1 =', v3_s_ap1 - v4_s_ap1)

#vibo: cумма членов рифметической прогрессии 24, 54 (шаг +30)
#vibo: первый член
ap2_a1 = 24
#vibo: последний член
ap2_an = 999999984
#vibo: количество членов арифметической прогрессии
ap2_n = ((ap2_an - ap2_a1)/30) + 1
#vibo: сумма членов армифметической прогрессии
v4_s_ap2 = (ap2_a1 + ap2_an) * ap2_n / 2
print(v4_s_ap2)
print('Проверка v3_s_ap2 - v4_s_ap2 =', v3_s_ap2 - v4_s_ap2)

#vibo: сумма членов арифметической прогрессии 27, 57 (шаг +30)
#vibo: первый член
ap3_a1 = 27
#vibo: последний член
ap3_an = 999999987
#vibo: количество членов арифметической прогрессии
ap3_n = ((ap3_an - ap3_a1)/30) + 1
#vibo: сумма членов армифметической прогрессии
v4_s_ap3 = (ap3_a1 + ap3_an) * ap3_n / 2
print(v4_s_ap3)
print('Проверка v3_s_ap3 - v4_s_ap3 =', v3_s_ap3 - v4_s_ap3)

print(v4_s_ap1 - v4_s_ap2 - v4_s_ap3)
print('Проверка:', (v4_s_ap1 - v4_s_ap2 - v4_s_ap3) - (v3_s_ap1 - v3_s_ap2 - v3_s_ap3))

print('Ответ:', int(v4_s_ap1 - v4_s_ap2 - v4_s_ap3))

1.6666666783333334e+17
Проверка v3_s_ap1 - v4_s_ap1 = 0.0
1.6666666633333332e+16
Проверка v3_s_ap2 - v4_s_ap2 = 0.0
1.6666666733333332e+16
Проверка v3_s_ap3 - v4_s_ap3 = 0.0
1.3333333446666667e+17
Проверка: 0.0
Ответ: 133333334466666672
CPU times: user 3.97 ms, sys: 0 ns, total: 3.97 ms
Wall time: 2.89 ms


#### Итоговое решение

In [1]:
%%time
#vibo: вариант 3 списки и идея арифметической прогрессии

#vibo: задаем список кратный трем, суммируем члены
v3_s_ap1 = sum(i for i in range(3, 1000000003, 3))
print(v3_s_ap1)

#vibo: задаем список 24, 54, и т.д. с шагом +30, суммируем члены
v3_s_ap2 = sum(i for i in range(24, 1000000003, 30))
print(v3_s_ap2)

#vibo: задаем список 27, 57 и т.д. с шагом +30, суммируем члены
v3_s_ap3 = sum(i for i in range(27, 1000000003, 30))
print(v3_s_ap3)

#vibo: из суммы членов первой арифметической прогрессии вычитаем суммы двух других
print('Ответ', v3_s_ap1 - v3_s_ap2 - v3_s_ap3)

166666667833333335
16666666633333332
16666666733333331
Ответ 133333334466666672
CPU times: user 24 s, sys: 8.05 ms, total: 24 s
Wall time: 24.2 s


##### Код для ods.ai

Сюда скопируйте и отправьте код, который решает задачу пунктом выше

In [2]:
v3_s_ap1 = sum(i for i in range(3, 1000000003, 3))
v3_s_ap2 = sum(i for i in range(24, 1000000003, 30))
v3_s_ap3 = sum(i for i in range(27, 1000000003, 30))
print(v3_s_ap1 - v3_s_ap2 - v3_s_ap3)

133333334466666672


### Вопрос 2.

На вход поступает текстовый файл из 3-х тысяч строк

Формат файла:
    "арифметическая операция"    "целое число #1"    "целое число #2"
Разделитель - 4 пробела


Нужно подготовить текстовый файл из 1 строки.
Строка содержит набор из 3-х тысяч чисел, разделенных запятой.
После последнего числа запятая не ставится.
каждое число - результат операции:
    "результирующее целое число" = "целое число #1" применить "арифметическая операция" "целое число #2"

#### Подготовка

In [1]:
#vibo: импортируем библиотеку
import operator
result = {'+': operator.add, '-': operator.sub, '*': operator.mul, '//': operator.floordiv, '%': operator.mod, '**': operator.pow}

In [2]:
#vibo: делаем проверку
a = 3
b = 2
print(result['+'](a, b), '=', a + b)
print(result['-'](a, b), '=', a - b)
print(result['*'](a, b), '=', a * b)
print(result['//'](a, b), '=', a // b)
print(result['%'](a, b), '=', a % b)
print(result['**'](a, b), '=', a ** b)

5 = 5
1 = 1
6 = 6
1 = 1
1 = 1
9 = 9


In [3]:
#vibo: на вход поступает строка
one_line = str('+    2    8')
#vibo: разделитель - четыре пробела
line_split = one_line.split('    ')
#vibo: проверяем
print(line_split)
#vibo: итоговая формула, проверяем
result[line_split[0]](int(line_split[1]),int(line_split[2]))

['+', '2', '8']


10

In [4]:
#vibo: импортируем библиотеку
import operator
result = {'+': operator.add, '-': operator.sub, '*': operator.mul, '//': operator.floordiv, '%': operator.mod, '**': operator.pow}
big_line = []
for one_line in open('test_input_file_1.txt', 'r').readlines():
    line_split = one_line.split('    ')
    one_answer = result[line_split[0]](int(line_split[1]),int(line_split[2]))
    big_line.append(one_answer)
print(one_answer)

508021860739623365322188197652216501772434524836001


In [5]:
#vibo: проверяем элементы списка
print(type(big_line))
print(big_line[0:10])

<class 'list'>
[-1551, 0, -49826439, 318, 68592064, -366, -1020, 1, 0, 5927]


In [6]:
#vibo: преобразуем каждый элемент list в str и склеиваем join
big_str = ",".join([str(_) for _ in big_line])

In [7]:
#vibo: посмотрим, что получилось
print(type(big_str))
print(big_str)

<class 'str'>
-1551,0,-49826439,318,68592064,-366,-1020,1,0,5927,14076019706120526112710656,4007885,349,-6084,1,5510,7342,-2698,1,372,-541,30431804,15694,59604644775390625,16835031,0,793714773254144,11026381951494519544538618432455883057857446928297348599765724730351258025260434645737117862538664702513136890902785802597870420145626388338001806739753231769,-52699853,1065,451,0,2623,0,8470329472543003390683225006796419620513916015625,2210,6149,-12142,2417851639229258349412352,1459,0,0,1286,70,2787015,-2848,-10033,154742504910672534362390528,0,-13255,12012,-16279,6253,-1363,0,-3710,981,6092,23605998,-9460,59768263894155949306790119265585619217025149412430681649,-2717526,6344,1,380,7,69690070456714036934135773061833984669532003071343585548161,13620860,37,7597,-7350,-13701,610,7922816251426433759354395033600000000000000000000000000000000,564,1,936,0,22,7425803930678705826362025654596757993811865030054903808,12167,-12139,18661,-32220474,-7403,-1340,-9376307,9369,1792,983,8736651,547150826329

In [8]:
#vibo: записываем ответ в файл
with open('v_test_output_file_1.txt', 'w') as f:
    f.write(big_str)

In [11]:
#vibo: делеаем проверку
#vibo: читаем данные тестового аутпута
test_output_file_1 = open ('test_output_file_1.txt', 'r')
test_output_object_1 = test_output_file_1.read()
#vibo: читаем данные моего аутпута
v_test_output_file_1 = open ('v_test_output_file_1.txt', 'r')
v_test_output_object_1 = v_test_output_file_1.read()
#vibo: сравнение двух str
test_output_object_1 == v_test_output_object_1

True

#### Итоговое решение

In [12]:
#vibo: теперь тоже самое для заданного файла

#vibo: импортируем библиотеку
import operator
result = {'+': operator.add, '-': operator.sub, '*': operator.mul, '//': operator.floordiv, '%': operator.mod, '**': operator.pow}
big_line = []
for one_line in open('input_file.txt', 'r').readlines():
    line_split = one_line.split('    ')
    one_answer = result[line_split[0]](int(line_split[1]),int(line_split[2]))
    big_line.append(one_answer)

#vibo: преобразуем каждый элемент list в str и склеиваем join
big_str = ",".join([str(_) for _ in big_line])

#vibo: записываем ответ в файл
with open('v_output_file_1.txt', 'w') as f:
    f.write(big_str)

##### Код для ods.ai

Сюда скопируйте и отправьте код, который решает задачу пунктом выше

In [2]:
import operator
result = {'+': operator.add, '-': operator.sub, '*': operator.mul, '//': operator.floordiv, '%': operator.mod, '**': operator.pow}
big_line = []
for one_line in open('input_file.txt', 'r').readlines():
    line_split = one_line.split('    ')
    one_answer = result[line_split[0]](int(line_split[1]),int(line_split[2]))
    big_line.append(one_answer)
big_str = ",".join([str(_) for _ in big_line])
with open('v_output_file_1.txt', 'w') as f:
    f.write(big_str)

### Вопрос 3.

На вход поступает два текстовый файл из 3-х тысяч строк каждый.

Первый файл содержит строки текста.
Второй файл содержит строки из двух целых неотрицательных чисел.

Первое число в строке всегда меньше или равно второму.
Числа всегда меньше длины соответствующей строки первого файла.
Соответствующей - это значит 1-ая строка из 1-го файла соответствует 1-ой строке из 2-го файла, а 123-я строка из 1-го файла соответствует 123-ей строке из 2-го файла.

Подготовить выходной файл, который состоит из подстрок 1-го входного файла. Подстроки разделены пробелами. Какие брать подстроки - написано во втором файле. В конце файла пробела нет.


#### Подготовка

In [39]:
#vibo: на вход подается строка из файла 1
one_str_f1 = 'XESADKaIFRctwwLIptQeRaKdDiNFdIHbvkanQUhtPGRIXNprSjXSagoRUzAThRVSHNxksFZkJmZUZYNLVuVumbEJynIBSCcjIXoBe'
print(type(one_str_f1))
#vibo: преобразуем строку к листу для индексации по данным из файла 2
one_list_f1 = list(one_str_f1)
#vibo: индексация по символам от 39 до 46 включая оба (правый край не входит, поэтому +1 индекс)
print(one_list[39:46+1])
#vibo: склеиваем результат слайса первой строки
join_answer_1 = "".join([str(_) for _ in one_list[39:46+1]])
print(join_answer_1)

<class 'str'>
['t', 'P', 'G', 'R', 'I', 'X', 'N', 'p']
tPGRIXNp


In [37]:
#vibo: на вход подается строка и из файла 2
one_str_f2 = '39 46'
print(type(one_str_f2))
#vibo: разрезаем по пробелу
one_str_f2_split = one_str_f2.split(' ')
print(one_str_f2_split)
#vibo: после метода join получили лист
print(type(one_str_f2_split))
#vibo: индексация НЕ ЗАБЫТЬ ПРО +1
print(one_str_f2_split[0])
print(one_str_f2_split[1])

<class 'str'>
['39', '46']
<class 'list'>
39
46


In [48]:
#vibo: теперь все тожже, но с индексацией
#vibo: на вход подается строка из файла 1
one_str_f1 = 'XESADKaIFRctwwLIptQeRaKdDiNFdIHbvkanQUhtPGRIXNprSjXSagoRUzAThRVSHNxksFZkJmZUZYNLVuVumbEJynIBSCcjIXoBe'
#vibo: преобразуем строку к листу для индексации по данным из файла 2
one_list_f1 = list(one_str_f1)

#vibo: на вход подается строка и из файла 2
one_str_f2 = '39 46'
#vibo: разрезаем по пробелу
one_str_f2_split = one_str_f2.split(' ')

#vibo: склеиваем результат слайса первой строки (правая граница +1)
join_answer_1 = "".join([str(_) for _ in one_list_f1[int(one_str_f2_split[0]):int(one_str_f2_split[1])+1]])
print(join_answer_1)

tPGRIXNp


In [15]:
#vibo: создаем пустой лист для записи результатов
big_line_2 = []
#vibo: на вход подается строка из файла 1 и файла 2, используем zip
for one_str_f1, one_str_f2 in zip(open('test_import_file_2_1.txt', 'r').readlines(), open('test_import_file_2_2.txt', 'r').readlines()):
    #vibo: преобразуем строку к листу для индексации по данным из файла 2
    one_list_f1 = list(one_str_f1)
    #vibo: разрезаем по пробелу cтроку из файла 2, чтобы взять индексы
    one_str_f2_split = one_str_f2.split(' ')
    #vibo: склеиваем результат слайса строки (правая граница +1)
    join_answer = "".join([str(_) for _ in one_list_f1[int(one_str_f2_split[0]):int(one_str_f2_split[1])+1]])
    #vibo: добавляем результат строки в лист
    big_line_2.append(join_answer)

#vibo: результат склеиваем в одну строку по условию задачи
big_str_2 = " ".join([str(_) for _ in big_line_2])
#vibo: записываем результат в файл
with open('v_test_output_file_2.txt', 'w') as f:
    f.write(big_str_2)

In [16]:
#vibo: делеаем проверку
#vibo: читаем данные тестового аутпута
test_output_file_2 = open ('test_output_file_2.txt', 'r')
test_output_object_2 = test_output_file_2.read()
#vibo: читаем данные моего аутпута
v_test_output_file_2 = open ('v_test_output_file_2.txt', 'r')
v_test_output_object_2 = v_test_output_file_2.read()
#vibo: сравнение двух str
test_output_object_2 == v_test_output_object_2

True

#### Итоговое решение

In [1]:
#vibo: теперь тоже самое для загрузки на платформу
#vibo: создаем пустой лист для записи результатов
big_line_2 = []
#vibo: на вход подается строка из файла 1 и файла 2, используем zip
for one_str_f1, one_str_f2 in zip(open('import_file_2_1.txt', 'r').readlines(), open('import_file_2_2.txt', 'r').readlines()):
    #vibo: преобразуем строку к листу для индексации по данным из файла 2
    one_list_f1 = list(one_str_f1)
    #vibo: разрезаем по пробелу cтроку из файла 2, чтобы взять индексы
    one_str_f2_split = one_str_f2.split(' ')
    #vibo: склеиваем результат слайса строки (правая граница +1)
    join_answer = "".join([str(_) for _ in one_list_f1[int(one_str_f2_split[0]):int(one_str_f2_split[1])+1]])
    #vibo: добавляем результат строки в лист
    big_line_2.append(join_answer)

#vibo: результат склеиваем в одну строку по условию задачи
big_str_2 = " ".join([str(_) for _ in big_line_2])
#vibo: записываем результат в файл
with open('v_output_file_2.txt', 'w') as f:
    f.write(big_str_2)

##### Код для ods.ai

Сюда скопируйте и отправьте код, который решает задачу пунктом выше

In [4]:
big_line_2 = []
for one_str_f1, one_str_f2 in zip(open('import_file_2_1.txt', 'r').readlines(), open('import_file_2_2.txt', 'r').readlines()):
    one_list_f1 = list(one_str_f1)
    one_str_f2_split = one_str_f2.split(' ')
    join_answer = "".join([str(_) for _ in one_list_f1[int(one_str_f2_split[0]):int(one_str_f2_split[1])+1]])
    big_line_2.append(join_answer)

big_str_2 = " ".join([str(_) for _ in big_line_2])
with open('v_output_file_2.txt', 'w') as f:
    f.write(big_str_2)

### Вопрос 4.

На вход поступает строка.
В ней хранится набор химических символов (He, O, H, Mg, Fe, ...). Без пробелов.
Нужно расшифровать химические символы в название химических элементов.
Для удобства, прилагается json файл, который ставит в соответствие химическому символу его химическое название.
Как прочитать этот файл в словарь python (dict):

periodic_table = json.load(open('periodic_table.json'))

Обратите внимание, что, например, "Bi" - это "Висмут", а не "БорЙод".
**То есть регистр (заглавные или прописные) букв имеет значение.**

#### Подготовка

In [7]:
#vibo: импортирируем
import json
periodic_table = json.load(open('periodic_table.json'))

In [9]:
print(type(periodic_table))
print(periodic_table)

<class 'dict'>
{'Ac': 'Актиний', 'Ag': 'Серебро', 'Al': 'Алюминий', 'Am': 'Америций', 'Ar': 'Аргон', 'As': 'Мышьяк', 'At': 'Астат', 'Au': 'Золото', 'B': 'Бор', 'Ba': 'Барий', 'Be': 'Бериллий', 'Bh': 'Борий', 'Bi': 'Висмут', 'Bk': 'Берклий', 'Br': 'Бром', 'C': 'Углерод', 'Ca': 'Кальций', 'Cd': 'Кадмий', 'Ce': 'Церий', 'Cf': 'Калифорний', 'Cl': 'Хлор', 'Cm': 'Кюрий', 'Cn': 'Коперниций', 'Co': 'Кобальт', 'Cr': 'Хром', 'Cs': 'Цезий', 'Cu': 'Медь', 'Db': 'Дубний', 'Ds': 'Дармштадтий', 'Dy': 'Диспрозий', 'Er': 'Эрбий', 'Es': 'Эйнштейний', 'Eu': 'Европий', 'F': 'Фтор', 'Fe': 'Железо', 'Fl': 'Флеровий', 'Fm': 'Фермий', 'Fr': 'Франций', 'Ga': 'Галий', 'Gd': 'Гадолиний', 'Ge': 'Германий', 'H': 'Водород', 'He': 'Гелий', 'Hf': 'Гафний', 'Hg': 'Ртуть', 'Ho': 'Гольмий', 'Hs': 'Хассий', 'I': 'Йод', 'In': 'Индий', 'Ir': 'Иридий', 'K': 'Калий', 'Kr': 'Криптон', 'La': 'Лантан', 'Li': 'Литий', 'Lr': 'Лоуренсий', 'Lu': 'Лютеций', 'Lv': 'Ливерморий', 'Mc': 'Московий', 'Md': 'Менделевий', 'Mg': 'Магний', 'M

In [46]:
#vibo: разбираем пример
import re
example_4 = 'NOTiFICaTiiiiiiON'
example_4 = re.findall(r'[A-Z]?[^A-Z]*', example_4)

example_4 = example_4[:int(len(example_4))-1]
print(example_4)

['N', 'O', 'Ti', 'F', 'I', 'Ca', 'Tiiiiii', 'O', 'N']


In [25]:
#vibo: загружаем библиотеки
import re
import json

#vibo: читаем данные из заданного файла
import_file_3 = open ('import_file_3.txt', 'r')
import_object_3 = import_file_3.read()

#vibo: разбиваем строку по буквам с верхним регистром
import_list_3 = re.findall(r'[A-Z]?[^A-Z]*', import_object_3)
#vibo: удаляем последний элемент списка (почему-то появлялся - '')
import_list_3 = import_list_3[:int(len(import_list_3))-1]

#vibo: импортирируем словарь
periodic_table = json.load(open('periodic_table.json'))

In [33]:
#vibo: лист из задания, который был в файле
import_list_3

['Cr',
 'Co',
 'Ra',
 'Li',
 'Bh',
 'Mn',
 'Md',
 'Sm',
 'Nh',
 'Pb',
 'Ca',
 'U',
 'U',
 'Mo',
 'Ir',
 'Po',
 'Eu',
 'C',
 'Pm',
 'Po',
 'Rb',
 'Pu',
 'Os',
 'Lr',
 'Ar',
 'Cm',
 'Pd',
 'Ce',
 'N',
 'Bk',
 'Ba',
 'La',
 'Cu',
 'As',
 'Cf',
 'Pr',
 'Cd',
 'Ba',
 'Nd',
 'Al',
 'Al',
 'Ru',
 'Ba',
 'Tl',
 'Fl',
 'Uue',
 'Ts',
 'Rn',
 'Ce',
 'Cr',
 'Hf',
 'Pm',
 'La',
 'Pb',
 'Md',
 'Cf',
 'Br',
 'Bk',
 'Pr',
 'Gd',
 'Es',
 'U',
 'Cf',
 'Ne',
 'Mg',
 'Sm',
 'La',
 'Bh',
 'Lu',
 'Ds',
 'Mo',
 'Sm',
 'Al',
 'O',
 'Pb',
 'Db',
 'Cu',
 'Ce',
 'Es',
 'Es',
 'Xe',
 'Ca',
 'Md',
 'He',
 'Sb',
 'Lv',
 'Th',
 'At',
 'F',
 'Ru',
 'Nb',
 'Pr',
 'Li',
 'Cu',
 'La',
 'Ni',
 'Yb',
 'Md',
 'Cl',
 'Ar',
 'Pm',
 'Ac',
 'Tm',
 'Gd',
 'At',
 'Lv',
 'Sr',
 'Nd',
 'Sc',
 'Tm',
 'F',
 'Mg',
 'Uue',
 'Ts',
 'Pd',
 'Br',
 'Sm',
 'Ru',
 'O',
 'Te',
 'Ac',
 'Lu',
 'Hs',
 'Tm',
 'Uue',
 'Yb',
 'Mn',
 'C',
 'Hf',
 'Dy',
 'Pb',
 'Ra',
 'Se',
 'Pm',
 'Hf',
 'Al',
 'K',
 'N',
 'Y',
 'Hg',
 'Pr',
 'Xe',
 'W',
 'H',
 'Ts

In [43]:
len(import_list_3)

500

In [29]:
#vibo: словарь с заданными названиями химических элементов
periodic_table

{'Ac': 'Актиний',
 'Ag': 'Серебро',
 'Al': 'Алюминий',
 'Am': 'Америций',
 'Ar': 'Аргон',
 'As': 'Мышьяк',
 'At': 'Астат',
 'Au': 'Золото',
 'B': 'Бор',
 'Ba': 'Барий',
 'Be': 'Бериллий',
 'Bh': 'Борий',
 'Bi': 'Висмут',
 'Bk': 'Берклий',
 'Br': 'Бром',
 'C': 'Углерод',
 'Ca': 'Кальций',
 'Cd': 'Кадмий',
 'Ce': 'Церий',
 'Cf': 'Калифорний',
 'Cl': 'Хлор',
 'Cm': 'Кюрий',
 'Cn': 'Коперниций',
 'Co': 'Кобальт',
 'Cr': 'Хром',
 'Cs': 'Цезий',
 'Cu': 'Медь',
 'Db': 'Дубний',
 'Ds': 'Дармштадтий',
 'Dy': 'Диспрозий',
 'Er': 'Эрбий',
 'Es': 'Эйнштейний',
 'Eu': 'Европий',
 'F': 'Фтор',
 'Fe': 'Железо',
 'Fl': 'Флеровий',
 'Fm': 'Фермий',
 'Fr': 'Франций',
 'Ga': 'Галий',
 'Gd': 'Гадолиний',
 'Ge': 'Германий',
 'H': 'Водород',
 'He': 'Гелий',
 'Hf': 'Гафний',
 'Hg': 'Ртуть',
 'Ho': 'Гольмий',
 'Hs': 'Хассий',
 'I': 'Йод',
 'In': 'Индий',
 'Ir': 'Иридий',
 'K': 'Калий',
 'Kr': 'Криптон',
 'La': 'Лантан',
 'Li': 'Литий',
 'Lr': 'Лоуренсий',
 'Lu': 'Лютеций',
 'Lv': 'Ливерморий',
 'Mc': 'Москови

In [41]:
#vibo: делаем пустой список для накопления результата
result_list = []
#vibo: идем по элементам списка из файла
for element in import_list_3:
    #vibo: иммем один элемент, проходим по ключам словаря
    for key in periodic_table:
        #vibo: если ключ словаря совпадает с элементом списка
        if key == element:
            print(key, ' = ', element, ' = ', periodic_table[key])
            #vibo: добавляем элемент в итоговый лист
            result_list.append(periodic_table[key])


Cr  =  Cr  =  Хром
Co  =  Co  =  Кобальт
Ra  =  Ra  =  Радий
Li  =  Li  =  Литий
Bh  =  Bh  =  Борий
Mn  =  Mn  =  Марганец
Md  =  Md  =  Менделевий
Sm  =  Sm  =  Самарий
Nh  =  Nh  =  Нихоний
Pb  =  Pb  =  Свинец
Ca  =  Ca  =  Кальций
U  =  U  =  Уран
U  =  U  =  Уран
Mo  =  Mo  =  Молибден
Ir  =  Ir  =  Иридий
Po  =  Po  =  Полоний
Eu  =  Eu  =  Европий
C  =  C  =  Углерод
Pm  =  Pm  =  Прометий
Po  =  Po  =  Полоний
Rb  =  Rb  =  Рубидий
Pu  =  Pu  =  Плутоний
Os  =  Os  =  Осмий
Lr  =  Lr  =  Лоуренсий
Ar  =  Ar  =  Аргон
Cm  =  Cm  =  Кюрий
Pd  =  Pd  =  Палладий
Ce  =  Ce  =  Церий
N  =  N  =  Азот
Bk  =  Bk  =  Берклий
Ba  =  Ba  =  Барий
La  =  La  =  Лантан
Cu  =  Cu  =  Медь
As  =  As  =  Мышьяк
Cf  =  Cf  =  Калифорний
Pr  =  Pr  =  Празеодим
Cd  =  Cd  =  Кадмий
Ba  =  Ba  =  Барий
Nd  =  Nd  =  Неодим
Al  =  Al  =  Алюминий
Al  =  Al  =  Алюминий
Ru  =  Ru  =  Рутений
Ba  =  Ba  =  Барий
Tl  =  Tl  =  Таллий
Fl  =  Fl  =  Флеровий
Uue  =  Uue  =  Унуненний
Ts  =  Ts  =  Те

In [42]:
#vibo:
print(result_list)

['Хром', 'Кобальт', 'Радий', 'Литий', 'Борий', 'Марганец', 'Менделевий', 'Самарий', 'Нихоний', 'Свинец', 'Кальций', 'Уран', 'Уран', 'Молибден', 'Иридий', 'Полоний', 'Европий', 'Углерод', 'Прометий', 'Полоний', 'Рубидий', 'Плутоний', 'Осмий', 'Лоуренсий', 'Аргон', 'Кюрий', 'Палладий', 'Церий', 'Азот', 'Берклий', 'Барий', 'Лантан', 'Медь', 'Мышьяк', 'Калифорний', 'Празеодим', 'Кадмий', 'Барий', 'Неодим', 'Алюминий', 'Алюминий', 'Рутений', 'Барий', 'Таллий', 'Флеровий', 'Унуненний', 'Теннессин', 'Радон', 'Церий', 'Хром', 'Гафний', 'Прометий', 'Лантан', 'Свинец', 'Менделевий', 'Калифорний', 'Бром', 'Берклий', 'Празеодим', 'Гадолиний', 'Эйнштейний', 'Уран', 'Калифорний', 'Неон', 'Магний', 'Самарий', 'Лантан', 'Борий', 'Лютеций', 'Дармштадтий', 'Молибден', 'Самарий', 'Алюминий', 'Кислород', 'Свинец', 'Дубний', 'Медь', 'Церий', 'Эйнштейний', 'Эйнштейний', 'Ксенон', 'Кальций', 'Менделевий', 'Гелий', 'Сурьма', 'Ливерморий', 'Торий', 'Астат', 'Фтор', 'Рутений', 'Ниобий', 'Празеодим', 'Литий', 'М

In [44]:
len(result_list)

500

In [51]:
#vibo: преобразуем каждый элемент list в str и склеиваем join
result_str = ''.join([str(_) for _ in result_list])
print(result_str)

ХромКобальтРадийЛитийБорийМарганецМенделевийСамарийНихонийСвинецКальцийУранУранМолибденИридийПолонийЕвропийУглеродПрометийПолонийРубидийПлутонийОсмийЛоуренсийАргонКюрийПалладийЦерийАзотБерклийБарийЛантанМедьМышьякКалифорнийПразеодимКадмийБарийНеодимАлюминийАлюминийРутенийБарийТаллийФлеровийУнуненнийТеннессинРадонЦерийХромГафнийПрометийЛантанСвинецМенделевийКалифорнийБромБерклийПразеодимГадолинийЭйнштейнийУранКалифорнийНеонМагнийСамарийЛантанБорийЛютецийДармштадтийМолибденСамарийАлюминийКислородСвинецДубнийМедьЦерийЭйнштейнийЭйнштейнийКсенонКальцийМенделевийГелийСурьмаЛиверморийТорийАстатФторРутенийНиобийПразеодимЛитийМедьЛантанНикельИттербийМенделевийХлорАргонПрометийАктинийТулийГадолинийАстатЛиверморийСтронцийНеодимСкандийТулийФторМагнийУнуненнийТеннессинПалладийБромСамарийРутенийКислородТеллурАктинийЛютецийХассийТулийУнуненнийИттербийМарганецУглеродГафнийДиспрозийСвинецРадийСеленПрометийГафнийАлюминийКалийАзотИттрийРтутьПразеодимКсенонВольфрамВодородТеннессинАстатХассийГерманийМышьяк

In [52]:
#vibo: записываем ответ в файл
with open('v_output_file_4.txt', 'w') as f:
    f.write(result_str)

#### Итоговое решение

In [2]:
#vibo: загружаем библиотеки
import re
import json

#vibo: читаем данные из заданного файла
import_file_3 = open ('import_file_3.txt', 'r')
import_object_3 = import_file_3.read()

#vibo: разбиваем строку по буквам с верхним регистром
import_list_3 = re.findall(r'[A-Z]?[^A-Z]*', import_object_3)
#vibo: удаляем последний элемент списка (почему-то появлялся - '')
import_list_3 = import_list_3[:int(len(import_list_3))-1]

#vibo: импортирируем словарь
periodic_table = json.load(open('periodic_table.json'))

#vibo: делаем пустой список для накопления результата
result_list = []
#vibo: идем по элементам списка из файла
for element in import_list_3:
    #vibo: иммем один элемент, проходим по ключам словаря
    for key in periodic_table:
        #vibo: если ключ словаря совпадает с элементом списка
        if key == element:
            #vibo: добавляем элемент в итоговый лист
            result_list.append(periodic_table[key])

#vibo: преобразуем каждый элемент list в str и склеиваем join
result_str = ''.join([str(_) for _ in result_list])

#vibo: записываем ответ в файл
with open('v_output_file_3.txt', 'w') as f:
    f.write(result_str)

##### Код для ods.ai

In [6]:
import re
import json

import_file_3 = open ('import_file_3.txt', 'r')
import_object_3 = import_file_3.read()

import_list_3 = re.findall(r'[A-Z]?[^A-Z]*', import_object_3)
import_list_3 = import_list_3[:int(len(import_list_3))-1]

periodic_table = json.load(open('periodic_table.json'))

result_list = []
for element in import_list_3:
    for key in periodic_table:
        if key == element:
            result_list.append(periodic_table[key])

result_str = ''.join([str(_) for _ in result_list])

with open('v_output_file_3.txt', 'w') as f:
    f.write(result_str)